toponetx/classes/path.py

```diff
@@ -0,0 +1,227 @@
+            if (
+                not reserve_sequence_order
+                and len(boundary) > 1
+                and str(boundary[0]) > str(boundary[-1])
+            ):
+                boundary.reverse()
+            if allowed_paths is None or tuple(boundary) in allowed_paths:
+                boundaries.append(tuple(boundary))
+        return boundaries
+
+    @property
+    def boundary(self) -> list[tuple[Hashable]]:
+        """
+        Return list of elementary (p-1)-path objects representing the elementary (p-1)-path on the boundary of the target elementary p-path.
+
+        Returns
+        -------
+        boundaries : list[tuple[Hashable]]
+            List of elementary p-path objects representing the boundary (p-1)-paths.
+        """
+        return self._boundaries
+
+    def clone(self) -> "Path":
+        """
+        Return a shallow copy of the elementary p-path.
+
+        Returns
+        -------
+        Path
+            A shallow copy of the elementary p-path.
+        """
+        return Path(
+            self.elements,
+            name=self.name,
+            construct_boundaries=self.construct_boundaries,
+            **self._properties,
+        )
+
+    def __check_inputs(self, elements: Any, reserve_sequence_order: bool):
+        """
+        Sanity check for inputs, as sequence order matters.
+
+        Parameters
+        ----------
+        elements: Any
+            The nodes in the elementary p-path.
+        reserve_sequence_order : bool
+            If True, reserve the order of the sub-sequence of nodes in the elementary p-path.
+            Else, the sub-sequence of nodes in the elementary p-path will be reversed if the first index is larger than the last index.
+            This variable is only used for sanity check on the input.
+        """
+        if isinstance(elements, Sequence):
+            if not isinstance(elements, list) and not isinstance(elements, tuple):
+                raise ValueError(
+                    f"Elements of an elementary p-path must be a list or tuple, got {type(elements)}"
+                )
+            for i in elements:
+                if not isinstance(i, Hashable):
+                    raise ValueError(f"All nodes of a p-path must be hashable, got {i}")
+            if (
+                not reserve_sequence_order
+                and len(elements) > 1
+                and str(elements[0]) > str(elements[-1])
+            ):
+                raise ValueError(
+                    "An elementary p-path must have the first index smaller than the last index, got {}".format(
+                        elements
+                    )
+                )
+        elif isinstance(elements, Hashable):
+            if isinstance(elements, int) or isinstance(elements, str):
+                pass
+            else:
+                raise ValueError(
+                    f"Elements of an elementary p-path must be a list or tuple, got {type(elements)}"
+                )
```

Comment:
```suggestion
            if not isinstance(elements, int) and not isinstance(elements, str):
                raise ValueError(
                    f"Elements of an elementary p-path must be a list or tuple, got {type(elements)}"
                )
```

Plus: The error message doesn't seem to correctly describe the problem here.

Reply:
Yes just replaced it with a more accurate error message.